pkg/controller/certificates/trigger/policies/gatherer_test.go

```diff
@@ -0,0 +1,336 @@
+			),
+			mockCertificateRequestsLister: mockCertificateRequests("default-unit-test-ns", func(t *testing.T) *listers.FakeCertificateRequestNamespaceLister {
+				f := expectCalled(t, 1)
+				return listers.NewFakeCertificateRequestNamespaceLister().WithList(func(_ labels.Selector) ([]*cmapi.CertificateRequest, error) {
+					f()
+					return nil, fmt.Errorf("error that is not a not_found error")
+				})
+			}),
+			wantErr: "error that is not a not_found error",
+		},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			g := &Gatherer{
+				CertificateRequestLister: test.mockCertificateRequestsLister(t),
+				SecretLister:             test.mockSecretLister,
+			}
+
+			got, gotErr := g.DataForCertificate(context.Background(), test.givenCert)
+
+			if test.wantErr != "" {
+				assert.Error(t, gotErr)
+				assert.EqualError(t, gotErr, test.wantErr)
+				return
+			}
+
+			require.NoError(t, gotErr)
+			assert.Equal(t, test.wantRequest, got.CurrentRevisionRequest)
+			assert.Equal(t, test.wantSecret, got.Secret)
+			assert.Equal(t, test.givenCert, got.Certificate, "input cert should always be equal to returned cert")
+		})
+	}
+}
+
+// Creates a mock CertificateRequestLister.
+//
+// We want to use a mock instead of a fake here: the mock makes sure that
+// (1) the lister.CertificateRequests(namespace) has been called with the
+// correct expected namespace and (2) lister.CertificateRequests(namespace)
+// has been called exactly once, which makes sure (1) was checked.
+func mockCertificateRequests(expNamespace string, innerLister func(t *testing.T) *listers.FakeCertificateRequestNamespaceLister) func(*testing.T) *listers.FakeCertificateRequestLister {
+	return func(t *testing.T) *listers.FakeCertificateRequestLister {
+		f := expectCalled(t, 1)
+		return listers.
+			NewFakeCertificateRequestLister().
+			WithCertificateRequests(func(namespace string) cmlist.CertificateRequestNamespaceLister {
+				f()
+				assert.Equal(t, expNamespace, namespace)
+				return innerLister(t)
+			})
+	}
+}
+
+// The returned function f is expected to be called expectedCount times.
+// Returns a friendly error with the location where f was created to help
+// the developer figure out where the error came from.
+func expectCalled(t *testing.T, expectedCount uint32) (f func()) {
+	// Debugging purposes: allows us to display where this function was
+	// called. The reason we need to to this is because t.Cleanup's stack
+	// does not contain the origin function, which makes it really hard to
+	// know where the assertion failed.
+	stack := grandparentStack(2)
+
+	gotCount := uint32(0)
+	t.Cleanup(func() {
+		assert.Equal(t, int(expectedCount), int(atomic.LoadUint32(&gotCount)), "this func was expected to be called %d times but was called %d times. Stack:\n\t%s",
+			expectedCount,
+			gotCount,
+			strings.Join(stack, "\n\t"),
+		)
+	})
+	return func() {
+		atomic.AddUint32(&gotCount, 1)
+	}
```

Comment:
It seems quite complicated to be introducing stacks and defers in testing clean up here. Why is it not sufficient to have an int which is checked and `t.Error`?

Reply:
I could not find a better way, how would you do it with `t.Error`?

Reply:
The `t.Error` will fall under the test case name.

Reply:
So:

```go
					t.Errorf("unexpected number of lister calls, exp=x got=%d", c.calls)
```

Reply:
I changed `assert.Equal` to `t.Error` (the `assert.Equal` was totally useless)

Not sure I understand the rest though; are you proposing a way to get rid of `t.Cleanup`?